react-common/components/theming/themeManager.ts

```diff
@@ -0,0 +1,121 @@
+/// <reference path='../../../localtypings/dompurify.d.ts' />
+
+interface ThemeChangeSubscriber {
+    subscriberId: string;
+    onColorThemeChange: () => void;
+}
+
+/*
+ * Class to handle theming operations, like loading and switching themes.
+ * Each instance can manage themes for a specific document.
+*/
+export class ThemeManager {
+    private static instances: Map<Document, ThemeManager> = new Map();
+    private currentTheme: pxt.ColorThemeInfo;
+    private subscribers: ThemeChangeSubscriber[];
+    private document: Document;
+
+    private constructor(doc: Document) {
+        this.document = doc;
+    }
+
+    public static getInstance(doc: Document = document): ThemeManager {
+        if (!ThemeManager.instances.has(doc)) {
+            ThemeManager.instances.set(doc, new ThemeManager(doc));
+        }
+        return ThemeManager.instances.get(doc);
+    }
+
+    public getCurrentColorTheme(): Readonly<pxt.ColorThemeInfo> {
+        return this.currentTheme;
+    }
+
+    public getAllColorThemes(): pxt.ColorThemeInfo[] {
+        return pxt.appTarget?.colorThemeMap ? Object.values(pxt.appTarget.colorThemeMap) : [];
+    }
+
+    // This is a workaround to ensure we still get all the special-case high-contrast styling
+    // until we fully support high contrast via color themes (requires a lot of overrides).
+    // TODO : this should be removed once we do fully support it.
+    private performHighContrastWorkaround(themeId: string) {
+        let setBodyClass = (add:boolean, className: string) => {
+            const body = this.document.body;
+            const hasClass = body.classList.contains(className);
```

Comment:
nitpick: no need to check contains here, classList.add already takes care of duplicates and classList.remove doesn't throw if missing